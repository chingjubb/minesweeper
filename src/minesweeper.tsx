import { Dispatch } from 'react';
import { Tile,
		 GameState,
		 GameAction,
		 ActionTypes } from './minesweeper_reducer';
import styles from './minesweeper.module.css';
import { Square } from './square';

type MinesweeperProps = {
	gameState: GameState;
	dispatch: Dispatch<GameAction>;
};

export const Minesweeper = ({ gameState, dispatch }: MinesweeperProps) => {
	const board: Tile[][] = gameState.board;
	const isAlive: boolean = gameState.isAlive;

	const renderBoard = () => {
		return <div>{board.map((thisRow: Tile[], row: number) => {
		return <div key={row} className={styles.row}>
				{thisRow.map((tile: Tile, column: number) => {
				return <Square {...tile}
							   key={`${row}.${column}.${tile.clicked}.${tile.flagged}`}
							   onRightClick={() => {
							   		dispatch({type: ActionTypes.setFlag,
							   				  row,
							   				  column})
							   }}
							   onClick={() => {
							   		dispatch({type: ActionTypes.clickTile,
							   				  row,
							   				  column})}
							}/>
					})}</div>
				})}</div>;
	};

	const renderGameOverText = () => {
		return (<div className={styles.gameOverText}
					onClick={() => {
						dispatch({ type: ActionTypes.restartGame }) }}>
					Game over! Click here to restart.
				</div>);
	};

	return (<div className={styles.gameBoard}>
				<div>{renderBoard()}</div>
				{!isAlive && renderGameOverText()}
			</div>);
};
