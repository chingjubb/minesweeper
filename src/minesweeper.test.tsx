import renderer from 'react-test-renderer';
import { Minesweeper } from './minesweeper';
import { Tile,
		 GameState,
		} from './minesweeper_reducer';

test('New game board', () => {
	const board: Tile[][] =
			 [[{row: 0, column: 0, isBomb: true, value: 0, clicked: false, flagged: false}],
			 [{row: 1, column: 0, isBomb: false, value: 1, clicked: false, flagged: false}]];

	const gameState: GameState = {
									board,
									isStarted: false,
									numBombs: 1,
									correctFlagged: 0,
									wrongFlagged: 0,
									isAlive: true };
 	const component = renderer.create(
		<Minesweeper gameState={gameState} dispatch={()=>{}} />,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Clicked a tile', () => {
	const board: Tile[][] =
		[[{row: 0, column: 0, isBomb: true, value: 0, clicked: false, flagged: false}],
		[{row: 1, column: 0, isBomb: false, value: 1, clicked: true, flagged: false}]];

	const gameState: GameState = {  board,
									isStarted: false,
									numBombs: 1,
									correctFlagged: 0,
									wrongFlagged: 0,
									isAlive: true };
 	const component = renderer.create(
		<Minesweeper gameState={gameState} dispatch={()=>{}} />,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Game over state', () => {
	const board: Tile[][] =
			 [[{row: 0, column: 0, isBomb: true, value: 0, clicked: true, flagged: false}],
			 [{row: 1, column: 0, isBomb: false, value: 1, clicked: false, flagged: false}]];

	const gameState: GameState = {
									board,
									isStarted: true,
									numBombs: 1,
									correctFlagged: 0,
									wrongFlagged: 1,
									isAlive: false };
 	const component = renderer.create(
		<Minesweeper gameState={gameState} dispatch={()=>{}} />,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Flag a bomb', () => {
	const board: Tile[][] =
			 [[{row: 0, column: 0, isBomb: true, value: 0, clicked: false, flagged: false}],
			 [{row: 1, column: 0, isBomb: false, value: 1, clicked: false, flagged: true}]];

	const gameState: GameState = {  board,
									isStarted: true,
									numBombs: 1,
									correctFlagged: 1,
									wrongFlagged: 1,
									isAlive: true };
 	const component = renderer.create(
		<Minesweeper gameState={gameState} dispatch={()=>{}} />,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('You won state', () => {
	const board: Tile[][] =
			 [[{row: 0, column: 0, isBomb: true, value: 0, clicked: false, flagged: true}],
			 [{row: 1, column: 0, isBomb: false, value: 1, clicked: false, flagged: false}]];

	const gameState: GameState = {  board,
									isStarted: true,
									numBombs: 1,
									correctFlagged: 1,
									wrongFlagged: 0,
									isAlive: true };
 	const component = renderer.create(
		<Minesweeper gameState={gameState} dispatch={()=>{}} />,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
