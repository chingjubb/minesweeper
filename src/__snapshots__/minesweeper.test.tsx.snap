// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clicked a tile 1`] = `
<div
  className="gameBoard"
>
  <div>
    <div>
      <div
        className="row"
      >
        <div
          className="tile unclickedTile"
          onClick={[Function]}
          onContextMenu={[Function]}
        />
      </div>
      <div
        className="row"
      >
        <div
          className="clickedTile tile oneBomb"
        >
          1
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Flag a bomb 1`] = `
<div
  className="gameBoard"
>
  <div>
    <div>
      <div
        className="row"
      >
        <div
          className="tile unclickedTile"
          onClick={[Function]}
          onContextMenu={[Function]}
        />
      </div>
      <div
        className="row"
      >
        <div
          className="tile flagged"
          onClick={[Function]}
          onContextMenu={[Function]}
        >
          ðŸš©
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Game over state 1`] = `
<div
  className="gameBoard"
>
  <div>
    <div>
      <div
        className="row"
      >
        <div
          className="bomb tile"
        >
          ðŸ’£
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="tile unclickedTile"
          onClick={[Function]}
          onContextMenu={[Function]}
        />
      </div>
    </div>
  </div>
  <div>
    <div>
      Game Over!
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Restart Game
      </span>
    </button>
  </div>
</div>
`;

exports[`New game board 1`] = `
<div
  className="gameBoard"
>
  <div>
    <div>
      <div
        className="row"
      >
        <div
          className="tile unclickedTile"
          onClick={[Function]}
          onContextMenu={[Function]}
        />
      </div>
      <div
        className="row"
      >
        <div
          className="tile unclickedTile"
          onClick={[Function]}
          onContextMenu={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`You won state 1`] = `
<div
  className="gameBoard"
>
  <div>
    <div>
      <div
        className="row"
      >
        <div
          className="tile flagged"
          onClick={[Function]}
          onContextMenu={[Function]}
        >
          ðŸš©
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="tile unclickedTile"
          onClick={[Function]}
          onContextMenu={[Function]}
        />
      </div>
    </div>
  </div>
  <div>
    <div>
      You Won!
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Restart Game
      </span>
    </button>
  </div>
</div>
`;
